### DISCUSSION TITLE
Noob Question: get state within async

### DISCUSSION THREAD

**Discussion by nyvs** [pmndrs/zustand#1852] [Q&A] [answered]
In this example login, setBasket, and fetchFruits work as expected.
fetchFruits sets a state member "fruits" so that I can log it to the console via an button onClick Listener no problem.

However, when I try to fetchColors, "fruits" is still null even when everything is awaited.
Calling the fetchColors function from outside via button onClick Listener works without flaw.

It must have something to do with async, but I cannot help myself.

```js
... // other store uses
const fruits = useFruitStore((state) => state.fruits);
const fetchFruits = useFruitStore((state) => state.fetchFruits);
const fetchColors = useColorStore((state) => state.fetchColors);
	
const handleLogin = async () => {
  await login("email@example.com", "1234");
  await setBasket("Basket I");
  await fetchFruits();
  await fetchColors(); // Doesn't work
};

const fetchColors = async () => {
  fruits.map(async (fruit) => { // Here, fruits is still null even when everything is awaited
    await fetchColorsByFruitRefer(fruit.uuid);
  });
};
```
I appreciate any help :)
Thanks

**Comment by dbritto-dev** [comment]
@nyvs It's ok to return a promise using map, without await it, then you can await the array of promises using `Promise.all` or `Promise.allSettled` like this

```ts
await Promise.all(fruits.map(fruit => fetchColorsByFruitRefer(fruit.uuid)))
```

let me know if that works for you

**Reply by nyvs** [reply]
Hi @dbritto-dev ! 
Do you mean like this?
I get the same error still :/

```js
const handleLogin = async () => {
  await login("email@example.com", "1234");
  await setBasket("Basket I");
  await fetchFruits();
  await fetchColors(); // Doesn't work
};

const fetchColors = async () => {
  // Error: Uncaught (in promise) TypeError: Cannot read properties of null (reading 'map')
  await Promise.all(fruits.map(fruit => fetchColorsByFruitRefer(fruit.uuid))) 
}
```

**Reply by dbritto-dev** [reply]
> Hi @dbritto-dev ! Do you mean like this? I get the same error still :/
> 
> ```js
> const handleLogin = async () => {
>   await login("email@example.com", "1234");
>   await setBasket("Basket I");
>   await fetchFruits();
>   await fetchColors(); // Doesn't work
> };
> 
> const fetchColors = async () => {
>   // Error: Uncaught (in promise) TypeError: Cannot read properties of null (reading 'map')
>   await Promise.all(fruits.map(fruit => fetchColorsByFruitRefer(fruit.uuid))) 
> }
> ```

That's a different error and that means `fruits` is null

**Reply by dbritto-dev** [reply]
@nyvs would you sharing us a basic example using codesandbox?

**Reply by nyvs** [reply]
yes that is my problem, `fruits` is null inside the async function, and is only populated after the function ends.
When I call setTimeout() for 10s inside `handleLogin` after `fetchFruits` I can run click fetchColors by clicking on a button, and it works. But I want to run it automatically inside `handleLogin` dirrectly after `fruits` is populated by `fetchFruits`...

**Reply by nyvs** [reply]
I can try making a codesandbox example, please give me some time @dbritto-dev

**Reply by nyvs** [reply]
@dbritto-dev 
https://codesandbox.io/s/unruffled-fast-9ndtjq?file=/src/App.js

**Reply by dbritto-dev** [reply, marked as answer]
@nyvs here you go https://codesandbox.io/s/wizardly-scooby-f29q2j?file=/src/App.js the main issue was fruits is outdated at that point, if you want to read the value after an async action you need to get it from the store itself using `useStore.getState()` or `get()`

**Reply by nyvs** [reply]
Thank you very much
What works best for me for now is getting the value via store action:
```js
getFruits: () => {
  return get().fruits
}
```
